File: src/main/java/com/mgjg/kmztracker/preference/MapPreferencesActivity.java
package com.mgjg.kmztracker.preference;

import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.preference.PreferenceFragment;

import com.mgjg.kmztracker.AppPreferences;
import com.mgjg.kmztracker.R;

public class MapPreferencesActivity extends PreferenceActivity
{

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        getFragmentManager().beginTransaction().replace(android.R.id.content, new MyPreferenceFragment()).commit();
    }

    public static class MyPreferenceFragment extends PreferenceFragment
    {
        @Override
        public void onCreate(final Bundle savedInstanceState)
        {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.preferences);

            final Preference pref = findPreference(getString(R.string.map_server_ip));

            final int key = R.string.map_server_preference;
            final String prefName = AppPreferences.getInstance().getIdentifier(key);
            String value = Preferences.getString(prefName, "");
            if (!value.isEmpty())
            {
                String summary = "current: " + value;
                pref.setSummary(summary);
                if (pref instanceof ListPreference)
                {
                    int ii = ((ListPreference) pref).findIndexOfValue(value);
                    ((ListPreference) pref).setValueIndex(ii);
                }
            }
            pref.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener()
            {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue)
                {
                    // do whatever you want with new value
                    Preferences.putString(prefName, String.valueOf(newValue));
                    // true to update the state of the Preference with the new value
                    // in case you want to disallow the change return false
                    return true;
                }
            });
        }
    }

}

File: src/main/java/com/mgjg/kmztracker/preference/MapPreferencesActivity.kt
package com.mgjg.kmztracker.preference

import android.os.Bundle
import android.preference.*
import com.mgjg.kmztracker.AppPreferences
import com.mgjg.kmztracker.R

class MapPreferencesActivity : PreferenceActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    fragmentManager.beginTransaction().replace(android.R.id.content, MyPreferenceFragment())
      .commit()
  }

  class MyPreferenceFragment : PreferenceFragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
      super.onCreate(savedInstanceState)
      addPreferencesFromResource(R.xml.preferences)

      val pref = findPreference(getString(R.string.map_server_ip))

      val key = R.string.map_server_preference
      val prefName = AppPreferences.instance!!.getIdentifier(key)
      var value = Preferences.getString(prefName, "")
      if (!value.isEmpty() && (pref is ListPreference)) {
          var ii = pref.findIndexOfValue(value)
          if (ii < 0) {
            ii = 0
            pref.setValueIndex(ii)
            value = pref.getValue()
            Preferences.putString(prefName, value)
          }
          val summary = "current: " + value
          pref.summary = summary
          pref.setValueIndex(ii)
      }
      pref.onPreferenceChangeListener =
          Preference.OnPreferenceChangeListener { preference, newValue ->
            val value = newValue.toString()
            Preferences.putString(prefName, value);
            if (pref is ListPreference) {
              val ii = pref.findIndexOfValue(value)
              // do whatever you want with new value
              // true to update the state of the Preference with the new value
              // in case you want to disallow the change return false
              val summary = "current: " + value
              pref.summary = summary
              pref.setValueIndex(ii)
            }
            true
          }
    }
  }

}

